### PR TITLE
Fixes #162

### LINKED ISSUE #162: Custom Placeholder Token Matching on optional token
https://github.com/auraphp/Aura.Router/blob/3.x/docs/defining-routes.md#custom-placeholder-token-matching

tweaking the supplied example so that date is optional (and even supplying a default date)

```php
<?php
$map->get('/calendar{/date}')
    ->tokens([
        'date' => function ($date, $route, $request) {
            try {
                new \DateTime($date);
                return true;
            } catch(\Exception $e) {
                return false;
            }
        }
    ])
    ->defaults(['date'=>'yesterday']);
```

will throw a notice:  `vendor/aura/router/src/Rule/Path.php (line 89):  = Undefined index: date`

notice is thrown even when supplying a default date value.

### PR DESCRIPTION
**Pull request by harikt**
We need to add a test for this.

I may not be able to find time for this.

@koriym can you look into this ?

### CODE REVIEW

File: src/Rule/Path.php [hunk 1]
```diff
@@ -86,7 +86,7 @@ public function __invoke(ServerRequestInterface $request, Route $route)
 
         foreach ($this->route->tokens as $name => $pattern) {
             if (is_callable($pattern)) {
-                if (!$pattern($attributes[$name], $route, $request)) {
+                if (!$pattern($attributes[$name] ?? NULL, $route, $request)) {
```
Comment (fisharebest) [inline src/Rule/Path.php]: For PSR-2, `NULL` should be `null`

https://www.php-fig.org/psr/psr-2/ - section 2.5

Comment (harikt) [reply src/Rule/Path.php]: Good catch.

**Comment by koriym** [comment]
@harikt @NaokiTsuchiya will take care of this. ARIGATO @NaokiTsuchiya

**Comment by harikt** [comment]
We may want to try if it is possible to merge 4.x to 3.x and support upto 8.1. if so we can start fresh for 8.1 only via 5.x branch and there by skip 4.x

**Comment by NaokiTsuchiya** [comment]
45c85e1ee2f258292e307ad1068c35abbfc32391 is a test for this PR.

How to add this commit to this PR ?

**Comment by harikt** [comment]
How about you send a different pr ?

**Comment by NaokiTsuchiya** [comment]
I have opened a PR that contains only tests. #188 
If you need a PR that contains the commits of this PR, I will create one.

**Comment by harikt** [comment]
I am merging this. Hope no issues. Thanks all.

**Pull request merged**